### DISCUSSION TITLE
stimp function

### DISCUSSION THREAD

**Discussion by hafarooki** [TDAmeritrade/stumpy#428] [Help: Coding & Implementations] [answered]
There is a `stumpy.stimp` function mentioned here: https://stumpy.readthedocs.io/en/latest/Tutorial_Pan_Matrix_Profile.html?highlight=pan%20#Compute-the-Pan-Matrix-Profile-using-STIMP

However, when I tried using it, it didn't seem to exist. Is there a beta version I need to use to access it? or a separate dependency?

**Comment by seanlaw** [comment, marked as answer]
@MicleBrick It is currently in beta and only available if you [install STUMPY directly from source](https://stumpy.readthedocs.io/en/latest/install.html#from-source). It should be released in the upcoming v1.9.0. Let me know if you need help or have any questions

**Reply by hafarooki** [reply]
So I tried adding this to my jupyter notebook:
```
!rm -rf stumpy_src
!git clone https://github.com/TDAmeritrade/stumpy.git stumpy_src
!conda install -c conda-forge -y numpy scipy numba
!cd stumpy_src && pip install .
```

```
import stumpy
```

It still said stumpy.stimp doesn't exist. I tried doing this:

```
from stumpy.stimp import stimp
```

...but I ended up with this error:

```
AttributeError: module 'stumpy.core' has no attribute 'get_max_window_size'
```

Eventually I tried this and it worked (stumpy_src is the folder I cloned the repo to):
```
import stumpy_src.stumpy as stumpy
```

So that worked although now I am unsure how to interpret this plot:
![image](https://user-images.githubusercontent.com/6323125/124489863-73e67400-dd7f-11eb-9c1e-297477867f5c.png)
I tried the `2 * n` thing to increase the resolution, but it's a perfect rectangle. Is that normal? What does it mean?

**Reply by seanlaw** [reply]
> So I tried adding this to my jupyter notebook:
> 
> ```
> !rm -rf stumpy_src
> !git clone https://github.com/TDAmeritrade/stumpy.git stumpy_src
> !conda install -c conda-forge -y numpy scipy numba
> !cd stumpy_src && pip install .
> ```
> 
> ```
> import stumpy
> ```
> 
> It still said stumpy.stimp doesn't exist. I tried doing this:
> 
> ```
> from stumpy.stimp import stimp
> ```
> 
> ...but I ended up with this error:
> 
> ```
> AttributeError: module 'stumpy.core' has no attribute 'get_max_window_size'
> ```
> 
> Eventually I tried this and it worked (stumpy_src is the folder I cloned the repo to):
> 
> ```
> import stumpy_src.stumpy as stumpy
> ```

@MicleBrick It looks like there are a lot of confounding factors here but the major issue is that it looks like you have multiple versions of STUMPY running (one installed via `conda` and another one installed from source). When you `import stumpy`, it is still using the `conda` version. When you are using `import stumpy_src` it is using a cloned version BUT this clone is sitting in the same directory as your Jupyter notebook, which is not the best way to get this installed properly. Right now, you are using the cloned package as a "Python module" and not as a "Python package". The best approach is to:

1. uninstall the conda version via `conda uninstall stumpy`
2. navigate to a separate directory (from your Jupyter notebook) and clone and install STUMPY as you did before so that you can import STUMPY as a Python package (i.e., `from stumpy.stimp import stimp`.

> 
> So that worked although now I am unsure how to interpret this plot:
> ![image](https://user-images.githubusercontent.com/6323125/124489863-73e67400-dd7f-11eb-9c1e-297477867f5c.png)
> I tried the `2 * n` thing to increase the resolution, but it's a perfect rectangle. Is that normal? What does it mean?

Firstly, I want to reiterate that pan matrix profiles computed from STIMP is an experimental function (so use at your own risk) but, in general, the pan matrix profile provides length-normalized matrix profile values with low values being black and high values being white. The pan matrix profile can sometimes be helpful in helping you figure out what an appropriate window size is. The "best" value of `m` is typically located at the peak(s) of a recurring right angle triangle (look for black triangles). There is something really strange in your plot, it looks like the colors are flipped and this is concerning as I do not know why that would be the case. The source of error could come from the data, manipulation of the results, or the plotting code. It is impossible for me to provide further support without more information.

Additionally, the current implementation of `stimp` only supports z-normalized distances.
